pylearn2/sandbox/nlp/models/mlp.py

```diff
@@ -189,3 +198,303 @@ def get_l1_weight_decay(self, coeff):
         assert isinstance(coeff, float) or hasattr(coeff, 'dtype')
         W, = self.transformer.get_params()
         return coeff * abs(W).sum()
+
+
+class ContrastiveProbabilityLayer(Layer):
+    """
```

Comment:
The distribution of positive and negative examples? Positive examples are taken from a dataset, negative ones are randomly generated. It can be as simple as a uniform distribution or corruption of positive examples. In Socher et al. (2013) they take a correct triple and replace the entity in either left hand side or right hand side. The replacement is constrained to be one that has appeared in the same relation and the same location.

Socher, R., Chen, D., Manning, C. D., & Ng, A. Y. (2013). Reasoning With Neural Tensor Networks for Knowledge Base Completion. Advances in Neural Information Processing Systems, 926–934.